sprint_10

Gvozdkov opened the pull request.

Commit:
Создан файл Constants и в нем реализованы  константы: AccessKey, SecretKey, RedirectURI, AccessScope и DefaultBaseURL.

Commit:
Исправлена реализация метода prepare в ImagesListViewController

Commit:
1. Создан AuthViewController, добавлен Image View и Button.
2, AuthViewController настроен согласно дизайну (https://www.figma.com/file/HyDfKh5UVPOhPZIhBqIm3q/Image-Feed-(YP)?type=design&node-id=334-4892&t=DjPogwfZoDqeh4Hp-0)

Commit:
Реализованно два контролера: Auth View Controller и Web View View Controller.
Написан запрос на получение кода авторизации в WebView.

Commit:
Используя KVO в WebViewViewController реализован процесс обновления загрузки веб-страницы.

Commit:
Создан стартовый SplashViewController контролер, который проверяет был ли ранее авторизован пользователь, в результате получения ответа отправляет пользователя или на флоу галереи или на флоу авторизации.

Commit:
Убран весь пишний закомментированный код

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,119 @@
+        urlComponents.queryItems = [
+            URLQueryItem(name: "client_id", value: AccessKey),
+            URLQueryItem(name: "redirect_uri", value: RedirectURI),
+            URLQueryItem(name: "response_type", value: "code"),
```
"code" тоже лучше в константы вынести

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,119 @@
+            URLQueryItem(name: "response_type", value: "code"),
+            URLQueryItem(name: "scope", value: AccessScope)
+        ]
+        let url = urlComponents.url!
```
Желательно не использовать force операции, лучше их избегать. Т.к они могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние.
Тут можно переписать с использование guard например

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,119 @@
+    private func updateProgress() {
+        progressView.progress = Float(webView.estimatedProgress)
+        progressView.isHidden = fabs(webView.estimatedProgress - 1.0) <= 0.0001
+        progressView.setProgress(1.0, animated: true)
```
Эта строчка у тебя убирает изменение прогресс бара, сразу устанавливает в полную загрузку и все.
Здорово, что ты нашел метод setProgress, он делает загрузку более плавной. Но тут должно быть тогда так:
```
         progressView.setProgress(Float(webView.estimatedProgress), animated: true)
         progressView.isHidden = fabs(webView.estimatedProgress - 1.0) <= 0.0001
```

helecta commented on ImageFeed/Constants.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+let AccessKey = "gHfk4CwFAJjd8aUgeyp6rtgehb7EHphx7gWB_jGHF2k"
```
Такие константы стоит называть со строчной буквы. Да, в учебнике было так, но так лучше не делать) С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,122 @@
+            return OAuth2TokenStorage().token
+        }
+        set {
+            OAuth2TokenStorage().token = newValue!
```
Тут тоже  force unwrapping, лучше переписать без него

helecta commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -0,0 +1,76 @@
+import UIKit
+//MARK: - SplashViewController
+final class SplashViewController: UIViewController {
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
```
Такие константы стоит называть со строчной буквы.

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,42 @@
+        if segue.identifier == ShowWebViewSegueIdentifier {
+            guard
+                let webViewViewController = segue.destination as? WebViewViewController
+            else { fatalError("Failed to prepare for \(ShowWebViewSegueIdentifier)") }
```
Обычно в продакшене не используют fatalError, советую заменить на assertionFailure, тут подробнее:
 [https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions](https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions)

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,42 @@
+
+//MARK: - class AuthViewController
+final class AuthViewController: UIViewController {
+    private let ShowWebViewSegueIdentifier = "ShowWebView"
```
Такие константы стоит называть со строчной буквы.

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,42 @@
+    weak var delegate: AuthViewControllerDelegate?
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == ShowWebViewSegueIdentifier {
```
Можно переписать короче:
```
        if segue.identifier == ShowWebViewSegueIdentifier,
           let webViewViewController = segue.destination as? WebViewViewController {
            
            webViewViewController.delegate = self
        }
        else {
            super.prepare(for: segue, sender: sender)
        }

```
Или с guard:
```
        guard segue.identifier == ShowWebViewSegueIdentifier,
              let webViewViewController = segue.destination as? WebViewViewController else {
            super.prepare(for: segue, sender: sender)
            return
        }
        webViewViewController.delegate = self
```
Здесь можно почитать про то [какие бенефиты дает нам guard](https://www.codingem.com/swift-guard/ )

helecta commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -0,0 +1,76 @@
+//MARK: - extension
+extension SplashViewController {
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == ShowAuthenticationScreenSegueIdentifier {
```
Тут тоже можно переписать короче. Аналогично как во AuthViewController

> Можно переписать короче:
> ```
>         if segue.identifier == ShowWebViewSegueIdentifier,
>            let webViewViewController = segue.destination as? WebViewViewController {
>             
>             webViewViewController.delegate = self
>         }
>         else {
>             super.prepare(for: segue, sender: sender)
>         }
> 
> ```
> Или с guard:
> ```
>         guard segue.identifier == ShowWebViewSegueIdentifier,
>               let webViewViewController = segue.destination as? WebViewViewController else {
>             super.prepare(for: segue, sender: sender)
>             return
>         }
>         webViewViewController.delegate = self
> ```
> Здесь можно почитать про то [какие бенефиты дает нам guard](https://www.codingem.com/swift-guard/ )

helecta commented on ImageFeed/SplashView/SplashViewController.swift:
```diff
@@ -0,0 +1,76 @@
+
+extension SplashViewController: AuthViewControllerDelegate {
+    func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String) {
+        dismiss(animated: true) { [weak self] in
```
У тебя получается, что сначала скрываешь контроллер, потом делаешь получение токена и потом переключение на основной экран.
Поэтому у тебя висит какое-то время опять экран авторизации, тк для SplashViewController успевает вызваться viewDidAppear.

Я бы сделала, что сначала получаем токен, и только потом скрываем экран:
```
oauth2Service.fetchAuthToken(code) { [weak self] result in
            dismiss(animated: true) { [weak self] in
                switch result {
                case .success:
                    self?.switchToTabBarController()
                case .failure:
                    break
                }
            }
        }
```

Commit:
Исправлено:
1. В ImageFeed/Auth/WebViewViewController.swift:
- у urlComponents.queryItems "code" вынесен константой,
- у let url = urlComponents.url! убран  force unrap,
- в  func updateProgress убрана лишняя строчка, которая устанавливает прогресс бара в полную загрузку.
2. В ImageFeed/Constants.swift - изменены названия всех констант
3. В ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
-  у var authToken убран  force unrap,
- у let ShowAuthenticationScreenSegueIdentifier заменено имя.
4. В ImageFeed/Auth/AuthViewController.swift:
- у func prepare убран fatalError,
- func prepare переписано в более короткий вариант и у let ShowWebViewSegueIdentifier изменено имя,
5. В ImageFeed/SplashView/SplashViewController.swift у extension SplashViewController изменена логика, где сначала получаем токен, а потом скрываем экран

Commit:
В  Main у Web View Controller изменен ключ класса на WebViewController

Commit:
Исправлена рекомендация frame у кнопки войти при отображение в Main
